pkg/controller/populators/upload-populator.go

```diff
@@ -0,0 +1,169 @@
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(uploadPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(uploadPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			installerLabels: installerLabels,
+			sourceKind:      cdiv1.VolumeUploadSourceRef,
+		},
+	}
+
+	uploadPopulator, err := controller.New(uploadPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	if err := addCommonPopulatorsWatches(mgr, uploadPopulator, log, cdiv1.VolumeUploadSourceRef, &cdiv1.VolumeUploadSource{}); err != nil {
+		return nil, err
+	}
+
+	return uploadPopulator, nil
+}
+
+// Reconcile the reconcile loop for the PVC with DataSourceRef of VolumeUploadSource kind
+func (r *UploadPopulatorReconciler) Reconcile(_ context.Context, req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("PVC", req.NamespacedName)
+	log.V(1).Info("reconciling Upload Source PVCs")
+	return r.reconcile(req, r, log)
+}
+
+// upload-specific implementation of getPopulationSource
+func (r *UploadPopulatorReconciler) getPopulationSource(namespace, name string) (client.Object, error) {
+	uploadSource := &cdiv1.VolumeUploadSource{}
+	uploadSourceKey := types.NamespacedName{Namespace: namespace, Name: name}
+	if err := r.client.Get(context.TODO(), uploadSourceKey, uploadSource); err != nil {
+		// reconcile will be triggered once created
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return uploadSource, nil
+}
+
+// Upload-specific implementation of updatePVCForPopulation
+func (r *UploadPopulatorReconciler) updatePVCForPopulation(pvc *corev1.PersistentVolumeClaim, source client.Object) {
+	pvc.Annotations[cc.AnnUploadRequest] = ""
+	uploadSource := source.(*cdiv1.VolumeUploadSource)
+	pvc.Annotations[cc.AnnContentType] = cc.GetContentType(string(uploadSource.Spec.ContentType))
+	pvc.Annotations[cc.AnnPopulatorKind] = cdiv1.VolumeUploadSourceRef
+}
+
+func (r *UploadPopulatorReconciler) updatePVCPrimeNameAnnotation(pvc *corev1.PersistentVolumeClaim, pvcPrimeName string) (bool, error) {
+	if _, ok := pvc.Annotations[AnnPVCPrimeName]; ok {
+		return false, nil
+	}
+
+	if pvc.Annotations == nil {
+		pvc.Annotations = make(map[string]string)
+	}
+	pvc.Annotations[AnnPVCPrimeName] = pvcPrimeName
+	if err := r.client.Update(context.TODO(), pvc); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (r *UploadPopulatorReconciler) removePVCPrimeNameAnnotation(pvc *corev1.PersistentVolumeClaim) error {
+	if _, ok := pvc.Annotations[AnnPVCPrimeName]; !ok {
+		return nil
+	}
+
+	delete(pvc.Annotations, AnnPVCPrimeName)
+	return r.client.Update(context.TODO(), pvc)
+}
+
+func (r *UploadPopulatorReconciler) reconcileTargetPVC(pvc, pvcPrime *corev1.PersistentVolumeClaim) (reconcile.Result, error) {
+	updated, err := r.updatePVCPrimeNameAnnotation(pvc, pvcPrime.Name)
+	if updated || err != nil {
```

Comment:
why return if updated?

Reply:
until its actually updated the upload cant start anyways, no really need to continue the reconcile loop